### PR TITLE
[WIP] Disallow deleting controller or isolated instance groups

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY
Added two new properties to the `InstanceGroup` model - `is_controller` and `is_isolated`. Used these properties to hide the trash icon for instance groups that are either controller or isolated.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

 - API
 - UI

##### AWX VERSION

```
awx: 4.0.0
```

### CODE REVIEW

**Comment by ghost** [comment]
@mabashian could you please take a look and add more on the UI side if necessary? The main task  was to make the condition for hiding trash icon stricter.

**Review by ryanpetrello — COMMENTED** [review]
It would be even better to also actually enforce this at the API level (and write tests that prove the enforcement).

File: awx/main/tests/functional/api/test_instance_group.py [hunk 1]
```diff
@@ -78,6 +78,20 @@ def instance_group_jobs_successful(instance_group, create_job_factory, create_pr
...
+    assert isolated_instance_group.is_isolated
+
+
+@pytest.mark.django_db
+def test_instance_group_is_controller(instance_group, isolated_instance_group):
+    # `instance_group` controls `isolated_instance_group`, hence a controller
+    assert instance_group.is_controller
+    assert not isolated_instance_group.is_controller
+
+
```
Comment (jladdjr) [inline awx/main/tests/functional/api/test_instance_group.py:100]: Maybe add a sanity check to make sure that empty instance groups aren't considered an isolated group or controller group?

File: awx/main/models/ha.py [hunk 2]
```diff
@@ -208,6 +208,15 @@ def jobs_running(self):
     def jobs_total(self):
         return UnifiedJob.objects.filter(instance_group=self).count()
 
+    @property
+    def is_controller(self):
+        return self.controlled_groups.exists()
+
+    @property
+    def is_isolated(self):
+        return all([instance.is_isolated for instance in self.instances.all()]) \
```
Comment (jladdjr) [inline awx/main/models/ha.py]: This looks like it should work. Checking all of the instances gave me pause though because TMK we don't allow mixed groups (that have both isolated and regular / full instances), so it seems like you could just do a check on the first instance you come across. I don't know that there's much of a penalty for querying all of the instances, though, so not sure that there's a big reason to change what's here. Just something that came to mind.

**Review by jladdjr — APPROVED** [review]
Added a couple of notes, but otherwise, LGTM! This seems really helpful to have in general in the API. Think @qe tests, in particular, could benefit by keying off of the new fields. Currently we're having to resort to queries like [this](https://github.com/jladdjr/tower-qa/blob/0090144bcb3d0d02fd90dbbbd54817e7f3b67ebe/tests/api/test_instances.py#L36).

File: awx/main/models/ha.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/models/ha.py]: This is making an unnecessary traversal. This is the InstanceGroup model, so the logic for this should be simply

```python
    @property
    def is_isolated(self):
        return bool(self.controller)
```

This information is already available to clients, so it's not necessary to provide this field for them to have all the information, but it's self-obvious to have a field named this.

File: awx/api/serializers.py [hunk 3]
```diff
@@ -4888,6 +4888,14 @@ class InstanceGroupSerializer(BaseSerializer):
         read_only=True
     )
     instances = serializers.SerializerMethodField()
+    is_controller = serializers.BooleanField(
+        help_text=_('Indicates whether instance group controls any other group'),
+        read_only=True
+    )
+    is_isolated = serializers.BooleanField(
+        help_text=_('Indicates whether all instance nodes of the group are isolated'),
```
Comment (AlanCoding) [inline awx/api/serializers.py]: I would say

> Indicates whether instances in this group are isolated. Isolated groups have a designated controller group.

Comment (jladdjr) [reply awx/api/serializers.py]: ^ good call. Isolated groups are an all or nothing thing. Either they're all isolated or all regular instances, you can't have a mixed group

**Comment by mabashian** [comment]
UI changes look good here

File: awx/main/tests/functional/api/test_instance_group.py (hunk 1 above)
Comment (jladdjr) [reply awx/main/tests/functional/api/test_instance_group.py:100]: What I wrote wasn't very clear. Basically, was thinking it'd be worth adding a test for the special case where an instance group is empty. If it was an isolated group (i.e. has a controller group) before it lost all of its nodes, it should still be an isolated group when empty. Similarly, if it was a regular instance group before all nodes were removed, then it should still be a regular instance group when empty.

File: awx/main/models/ha.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/models/ha.py]: this is still using unneeded complicated logic which may not be correct in all cases.

**Comment by ghost** [comment]
> It would be even better to also actually enforce this at the API level (and write tests that prove the enforcement).

@ryanpetrello don't we already have the enforcement in place?

https://github.com/ansible/awx/blob/devel/awx/api/views/__init__.py#L389-L392

**Comment by ryanpetrello** [comment]
@vismay-golwala we do, thanks for checking :)

**Comment by ryanpetrello** [comment]
recheck

File: awx/main/models/ha.py (hunk 2 above)
Comment (ghost) [reply awx/main/models/ha.py]: @AlanCoding marked as resolved a bit early, my bad

**Review by ryanpetrello — APPROVED** [review]
This looks good to me.

Let's hold off on merging this for just a bit.

File: awx/main/models/ha.py [hunk 4]
```diff
@@ -208,6 +208,14 @@ def jobs_running(self):
     def jobs_total(self):
         return UnifiedJob.objects.filter(instance_group=self).count()
 
+    @property
+    def is_controller(self):
+        return self.controlled_groups.exists()
```
Comment (jladdjr) [inline awx/main/models/ha.py]: 💯

File: awx/main/models/ha.py [hunk 5]
```diff
@@ -208,6 +208,14 @@ def jobs_running(self):
...
+
+    @property
+    def is_isolated(self):
+        return bool(self.controller)
```
Comment (jladdjr) [inline awx/main/models/ha.py:217]: 💯

**Comment by ghost** [comment]
Is this good to merge now?

**Comment by AlanCoding** [comment]
I think so. I find the `is_controller` field to be a little redundant, since the UI could just as well check `controller`, but there's no performance penalty for that. This certainly seems to do what's asked for.

**Comment by ghost** [comment]
> I think so. I find the `is_controller` field to be a little redundant, since the UI could just as well check `controller`, but there's no performance penalty for that. This certainly seems to do what's asked for.

Yeah we can definitely do that, but I think this way it is more readable in the UI part. When we try to comprehend it now, it makes more sense that we are hiding the trash icon because it `is_controller` rather than its `controlled_group.exists()` .
